gateway/gateway.go

```diff
@@ -13,27 +13,34 @@ var (
 	ErrNotValidHandshake = errors.New("this connection didnt provide a valid handshake")
 )
 
-func NewBasicGatewayWithStore(store ServerStore, ch <-chan connection.HandshakeConn) BasicGateway {
-	return BasicGateway{store: store, inCh: ch}
+func NewBasicGatewayWithStore(store ServerStore, connCh <-chan connection.HandshakeConn, closeCh <-chan struct{}) BasicGateway {
+	return BasicGateway{store: store, inCh: connCh, closeCh: closeCh}
 }
 
 type ServerData struct {
 	ConnCh chan<- connection.HandshakeConn
 }
 
 type BasicGateway struct {
-	store ServerStore
-	inCh  <-chan connection.HandshakeConn
+	store   ServerStore
+	inCh    <-chan connection.HandshakeConn
+	closeCh <-chan struct{}
 }
 
 func (g *BasicGateway) Start() error {
+Forloop:
 	for {
-		conn := <-g.inCh
-		g.HandleConn(conn)
+		select {
+		case conn := <-g.inCh:
+			g.handleConn(conn)
```

Comment:
Starting a goroutine here defeats the purpose of the worker pool, which is the current design we're using. 

If we were to start a goroutine here it would need to copy the all the data which is used in the gateway, now should that be such a big problem anymore since not every goroutine will now have the entire gateway with all proxy servers and their config values with the map which contains players names and their connections. 
Im also not sure why you wanna start a new goroutine for every connection, ik its a common pattern in rest api's but there you share some basic application wide context with each connection and the rest of the data you get either from the database or from the user self. 

There also are some other things you should probably be aware of. With select number of working gateway it would be easier for an attacker to try to occupy them all so nobody else can join them(this is why we should use deadlines/timeouts), but at the same time this also means that an attacker cant open thousands of connections and thus that many goroutines within a short amount of time. With less number of goroutines which are acting as gateways it should also be easier to build code which still runs fast but with an anti bot filter built in for example.  This is what I can think of right now but there probably are also other things

Reply:
We should implement a rate limiter or connection limit per IP to at least stop attackers from opening as many connections as they want. I don't know anything about bot filters yet, but I will do my research if I have time. My main concern is a [Slowloris](https://en.wikipedia.org/wiki/Slowloris_(computer_security)) attack. Just setting deadlines won't protect us from that.

Reply:
Yes and building such limiter is easier to do with when you have 20 goroutines then when you have 1k goroutines, its also easier to build some code which will drop existing connections. 
I also dont know much about bot filters but an easier implementation could be when its turned on that you save the player name the users uses to connect with you and than you disconnect them. If someone on the same ip is trying to connection with a different username you block that ip. 

Yes slowloris attack is a problem, but that goes for both ways as far as I see.  Like the current infrared version also when too many calls get through the gateway, there will be lots of connections being opened with which will cause the server problems. 

Also i just realized we could easily just make a different gateway which does open a goroutine for each connection there and processes it like that. We only have to an like a config value so that people can pick what they prefer themselves